src/common/procedure/src/store/util.rs

```diff
@@ -0,0 +1,273 @@
+use snafu::{ensure, ResultExt};
+
+use super::state_store::KeySet;
+use crate::error;
+use crate::error::Result;
+
+pub struct CollectingState {
+    pairs: Vec<(String, Vec<u8>)>,
+}
+
+fn parse_segments(segments: Vec<(String, Vec<u8>)>, prefix: &str) -> Result<Vec<(usize, Vec<u8>)>> {
+    segments
+        .into_iter()
+        .map(|(key, value)| {
+            let suffix = key.trim_start_matches(prefix);
+            let index = suffix
+                .parse::<usize>()
+                .context(error::ParseSegmentKeySnafu { key })?;
+
+            Ok((index, value))
+        })
+        .collect::<Result<Vec<_>>>()
+}
+
+/// Collects multiple values into a single key-value pair.
+/// Returns an error if:
+/// - Part values are lost.
+/// - Failed to parse the key of segment.
+pub fn multiple_values_collector(
+    CollectingState { mut pairs }: CollectingState,
+) -> Result<(KeySet, Vec<u8>)> {
+    if pairs.len() == 1 {
+        // Safety: must exist.
+        let (key, value) = pairs.into_iter().next().unwrap();
+        Ok((KeySet::new(key, 0), value))
+    } else {
+        let segments = pairs.split_off(1);
+        // Safety: must exist.
+        let (key, value) = pairs.into_iter().next().unwrap();
+        let prefix = KeySet::with_prefix(&key);
+        let mut parsed_segments = parse_segments(segments, &prefix)?;
+        parsed_segments.sort_unstable_by(|a, b| a.0.cmp(&b.0));
+
+        // Safety: `parsed_segments` must larger than 0.
+        let segment_num = parsed_segments.last().unwrap().0;
+        ensure!(
+            // The segment index start from 1.
+            parsed_segments.len() == segment_num,
+            error::UnexpectedSnafu {
+                err_msg: format!(
+                    "Corrupted segment keys, parsed segment indexes: {:?}",
+                    parsed_segments
+                        .into_iter()
+                        .map(|(key, _)| key)
+                        .collect::<Vec<_>>()
+                )
+            }
+        );
+
+        let segment_values = parsed_segments.into_iter().map(|(_, value)| value);
+        let mut values = vec![value];
+        values.extend(segment_values);
+
+        Ok((KeySet::new(key, segment_num), values.concat()))
+    }
+}
+
+impl CollectingState {
+    fn new(key: String, value: Vec<u8>) -> CollectingState {
+        Self {
+            pairs: vec![(key, value)],
+        }
+    }
+
+    fn push(&mut self, key: String, value: Vec<u8>) {
+        self.pairs.push((key, value));
+    }
+
+    fn key(&self) -> &str {
+        self.pairs[0].0.as_str()
+    }
+}
+
+#[derive(Debug)]
+enum MultipleValuesStreamState {
+    Idle,
+    Collecting,
+    End,
+}
+
+pub type Collector = dyn Fn(CollectingState) -> Result<(KeySet, Vec<u8>)> + Send;
+pub type Upstream = dyn Stream<Item = Result<(String, Vec<u8>)>> + Send;
+
+/// A stream collects multiple values into a single key-value pair.
+pub struct MultipleValuesStream {
+    upstream: Pin<Box<Upstream>>,
```

Comment:
Is it possible that the output ordering of `Upstream` is unspecific?

Reply:
No, it requires output ordering of `Upstream` in the lexical order.